source/common/http/conn_manager_impl.cc

```diff
@@ -466,6 +485,56 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
       connection_manager_.config_, *snapped_route_config_, connection_manager_.random_generator_,
       connection_manager_.runtime_, connection_manager_.local_info_);
 
+  if (!cached_route_.valid()) {
+    cached_route_.value(snapped_route_config_->route(*request_headers_, stream_id_));
+  }
+
+  // Check for WebSocket upgrade request if the route is valid and there
+  // are no pending responses except for the current one. This is not the
+  // expected behavior, as a client can pipeline several requests - the tail
+  // end of which can be a WebSocket upgrade. However, handling this edge scenario
+  // would require us to check for upgrades in encodeHeaders where the machinery
+  // gets lot more complicated. So, for the moment, when we have outstanding responses,
+  // we reject the upgrade request. From a browser point of view, Chrome/Mozilla disable
+  // pipelining, while IE11+ does not even support pipelining. So this limitation should
```

Comment:
Correct, we do not do real pipelining in the server since I did not want to deal with out of order responses for HTTP/1.1. We do allow the client to pipeline and handle that correctly. I don't think we will ever support full HTTP/1.1 pipelining in Envoy. At this point people can use HTTP/2 if they care. For HTTP/1.1 there will never be any pending responses. It will always be one request at a time until a response is sent.